main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -33,303 +33,518 @@ object Safety {
+      case (Instance(_, _, _, _, tpe, _, _, _, loc), _) =>
+        if (tpe.typeArguments.exists(_.kind == Kind.Bool)) {
+          SoftFailure[Root, CompilationMessage](root, LazyList(SafetyError.SendableError(tpe, loc)))
+        } else {
+          root.toSuccess[Root, CompilationMessage]
+        }
     }
   }
 
   /**
     * Performs safety and well-formedness checks on the given definition `def0`.
     */
-  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = {
     val renv = def0.spec.tparams.map(_.sym).foldLeft(RigidityEnv.empty) {
       case (acc, e) => acc.markRigid(e)
     }
-    visitExp(def0.impl.exp, renv, root)
+    mapN(visitExp(def0.impl.exp, renv, root))(_ => root)
   }
 
 
   /**
     * Performs safety and well-formedness checks on the given expression `exp0`.
     */
-  private def visitExp(e0: Expression, renv: RigidityEnv, root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitExp(e0: Expression, renv: RigidityEnv, root: Root)(implicit flix: Flix): Validation[Expression, CompilationMessage] = {
 
-    def visit(exp0: Expression): List[CompilationMessage] = exp0 match {
-      case Expression.Cst(_, _, _) => Nil
+    def visit(exp0: Expression): Validation[Expression, CompilationMessage] = exp0 match {
+      case Expression.Cst(_, _, _) => Success(exp0)
 
-      case Expression.Wild(_, _) => Nil
+      case Expression.Wild(_, _) => Success(exp0)
 
-      case Expression.Var(_, _, _) => Nil
+      case Expression.Var(_, _, _) => Success(exp0)
 
-      case Expression.Def(_, _, _) => Nil
+      case Expression.Def(_, _, _) => Success(exp0)
 
-      case Expression.Sig(_, _, _) => Nil
+      case Expression.Sig(_, _, _) => Success(exp0)
 
-      case Expression.Hole(_, _, _) => Nil
+      case Expression.Hole(_, _, _) => Success(exp0)
 
       case Expression.HoleWithExp(exp, _, _, _, _) =>
-        visit(exp)
+        for {_ <- visit(exp)
+             x <- Success(exp0)} yield x
 
       case Expression.Use(_, exp, _) =>
-        visit(exp)
+        for {_ <- visit(exp)
+             x <- Success(exp0)} yield x
 
       case Expression.Lambda(_, exp, _, _) =>
-        visit(exp)
+        for {_ <- visit(exp)
+             x <- Success(exp0)} yield x
 
       case Expression.Apply(exp, exps, _, _, _, _) =>
-        visit(exp) ::: exps.flatMap(visit)
+        for {
+          _ <- Validation.foldRight(exps)(visit(exp))((e, _) => visit(e))
+          x <- Success(exp0)
+        } yield x
 
       case Expression.Unary(_, exp, _, _, _, _) =>
-        visit(exp)
+        for {
+          _ <- visit(exp)
+          x <- Success(exp0)
+        } yield x
 
       case Expression.Binary(_, exp1, exp2, _, _, _, _) =>
-        visit(exp1) ::: visit(exp2)
+        for {
```

Comment:
We should always avoid `for` because this treats `Validation` as a Monad, but actually its a lie. We should only treat it as an applicative functor.

Reply:
Indeed, but if we have two sub-expressions we want to collect errors from both of them :)